### PR TITLE
fix: Add flag-protected informer write-back

### PR DESCRIPTION
**Pull request by simster7**
Closes https://github.com/argoproj/argo/issues/3719

I have tested some Workflows known to cause the "conflict" error using the strategy outlined in https://github.com/argoproj/argo/pull/3684#issue-464115275 and have so for not seen any issues when using the informer write-back strategy.

My current idea is that we merge this as an opt-in feature-flag protected change. Ask users to test, then consider making it a permanent change if there are no issues.

Things I'd like to consider before moving forward with this:

* [ ] Find and understand the rationale for why writing back to the informer is an anti pattern (talk to @jessesuen)
* [ ] Write a "stress test" to try to find any weak points with this strategy
* [ ] How much benefit we get from dropping the "sleep for 1s" strategy

### CODE REVIEW

File: workflow/controller/operator.go [hunk 1]
```diff
@@ -529,7 +530,22 @@ func (woc *wfOperationCtx) persistUpdates() {
...
+		woc.log.Infof("attempting informer write back")
+		un, err := wfutil.ToUnstructured(woc.wf)
+		if err != nil {
+			woc.log.Errorf("error converting workflow to unstructured: %s", err)
+			time.Sleep(1 * time.Second)
+		} else {
+			err = woc.controller.wfInformer.GetStore().Update(un)
+			if err != nil {
+				woc.log.Errorf("error writing back workflow to informer: %s", err)
+				time.Sleep(1 * time.Second)
```
Comment (simster7) [inline workflow/controller/operator.go:543]: If the write back fails, simply sleep as we normally would

Comment (alexec) [reply workflow/controller/operator.go:543]: I think this is a good idea.

File: workflow/controller/operator.go [hunk 2]
```diff
@@ -529,7 +530,22 @@ func (woc *wfOperationCtx) persistUpdates() {
 	// conflicts arose when attempting to update, causing our conflict resolution code to be invoked. The conflict
 	// resolution code was not perfect and workflow information was not correctly persisted, causing undefined behavior.
 	// This line was reintroduced in v2.9.5, and should remain as long as we use our current informer pattern.
-	time.Sleep(1 * time.Second)
+	if os.Getenv("INFORMER_WRITE_BACK") == "true" {
```
Comment (alexec) [inline workflow/controller/operator.go:533]: I think that having this off by default is a good idea. It has cons - we have an untested path of code, which mean it'll rot at some point.

If we can create a test that we can be confident if that we have fixed this, I'd suggest we flip this guard.

**Comment by alexec** [comment]
I wonder if we're approaching this incorrectly. Rather than finding a solution, should we be finding a reliable way to reproduce these kinds of issues? A chaos suite for example?

**Comment by alexec** [comment]
Closing as we have #4025

**Pull request closed without merging**